lib/python/rose/config.py

```diff
@@ -1378,38 +1381,47 @@ def load(self, source, node=None, default_comments=None):
-                    raise ConfigSyntaxError(
-                        ConfigSyntaxError.BAD_CHAR,
-                        file_name, line_num, len(head) + bad_index, line)
-                # Find position under root node
-                if type_ == self.TYPE_OPTION:
-                    keys.pop()
-                if keys:
-                    keys.pop()
-                section = section.strip()
-                if section:
-                    keys.append(section)
-                    type_ = self.TYPE_SECTION
-                else:
-                    keys = []
-                    type_ = None
-                section_node = node.get(keys[:])
-                if section_node is None:
-                    node.set(keys[:], {}, state, comments)
+                if self.allow_sections:
+                    head, section, state = match.group(
+                        "head", "section", "state")
+                    bad_index = self._check_section_value(section)
+                    if bad_index > -1:
+                        raise ConfigSyntaxError(
+                            ConfigSyntaxError.BAD_CHAR,
+                            file_name, line_num, len(head) + bad_index, line)
+                    # Find position under root node
+                    if type_ == self.TYPE_OPTION:
+                        keys.pop()
+                    if keys:
+                        keys.pop()
+                    section = section.strip()
+                    if section:
+                        keys.append(section)
+                        type_ = self.TYPE_SECTION
+                    else:
+                        keys = []
+                        type_ = None
+                    section_node = node.get(keys[:])
+                    if section_node is None:
+                        node.set(keys[:], {}, state, comments)
+                    else:
+                        section_node.state = state
+                        if comments:
+                            section_node.comments += comments
+                    comments = []
+                    continue
                 else:
-                    section_node.state = state
-                    if comments:
-                        section_node.comments += comments
-                comments = []
-                continue
+                    raise ConfigSyntaxError(
+                        ConfigSyntaxError.SECTIONS_NOT_ALLOWED, file_name,
+                        line_num, 0, line)
             # Match the start of an option setting?
             match = self.re_option.match(line)
             if not match:
-                raise ConfigSyntaxError(
-                    ConfigSyntaxError.BAD_SYNTAX, file_name, line_num, 0, line)
+                if self.allow_sections:
+                    err = ConfigSyntaxError.BAD_SYNTAX
+                else:
+                    err = ConfigSyntaxError.BAD_SYNTAX_NO_SECTIONS
+                raise ConfigSyntaxError(err, file_name, line_num, 0, line)
```

Comment:
Now that sections are handled separately this can go back the way it was?

```suggestion
                raise ConfigSyntaxError(
                      ConfigSyntaxError.BAD_SYNTAX, file_name, line_num, 0, line)
```